File: bin/my_scripts/utils/format_merge_lines.rb
require 'find'

def check_first_line(line)
    temp_line = line.strip
    (temp_line.empty? || temp_line.start_with?('/') || temp_line.start_with?('*')) && !temp_line.start_with?('#')
end

def check_for_comment(line)
    line.strip.start_with?('/') || line.strip.start_with?('*')
end

def check_for_content(line)
    !line.strip.empty? && !check_for_comment(line) && !line.strip.start_with?('#')
end

def check_for_paren(line)
    line.strip.include?('(')
end

def check_for_end_paren(line)
    line.strip.end_with?(')')
end

def check_for_func_start(line)
    line.gsub("\n", '').start_with?('{')
end

def process_file(filename)
    lines = File.readlines(filename)
    new_lines = []
    change_counter = 0

    i = 0
    while i < lines.size
        if i < lines.size - 3 && check_first_line(lines[i]) && check_for_content(lines[i + 1]) && check_for_end_paren(lines[i + 2]) && check_for_func_start(lines[i + 3])
            new_lines << lines[i]
            new_lines << lines[i + 1].strip + ' ' + lines[i + 2].strip
            new_lines << lines[i + 3]
            i += 4
            change_counter += 1
        else
            new_lines << lines[i]
            i += 1
        end
    end

    File.write(filename, new_lines.join("\n"))
    change_counter
end

dir_path = './' # Current directory
total_change_count = 0

Find.find(dir_path) do |path|
    next unless path.end_with?('.c')
    puts "Processing file: #{path}"
    file_change_count = process_file(path)
    puts "Done! Changes done: #{file_change_count}"
    total_change_count += file_change_count
end

puts "Done processing files! Total changes done: #{total_change_count}"

File: bin/my_scripts/utils/format_merge_lines.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <algorithm>
#include <filesystem>
namespace fs = std::filesystem;

bool checkFirstLine(const std::string& line) {
    std::string tempLine = line;
    tempLine.erase(std::remove_if(tempLine.begin(), tempLine.end(), isspace), tempLine.end());
    return (tempLine.empty() || tempLine[0] == '/' || tempLine[0] == '*') && tempLine[0] != '#';
}

bool checkForComment(const std::string& line) {
    std::string tempLine = line;
    tempLine.erase(std::remove_if(tempLine.begin(), tempLine.end(), isspace), tempLine.end());
    return tempLine[0] == '/' || tempLine[0] == '*';
}

bool checkForContent(const std::string& line) {
    std::string tempLine = line;
    tempLine.erase(std::remove_if(tempLine.begin(), tempLine.end(), isspace), tempLine.end());
    return !tempLine.empty() && !checkForComment(tempLine) && tempLine[0] != '#';
}

bool checkForParen(const std::string& line) {
    return line.find('(') != std::string::npos;
}

bool checkForEndParen(const std::string& line) {
    return line.find(')') != std::string::npos && line.find(')') == line.size() - 1;
}

bool checkForFuncStart(const std::string& line) {
    return line.find('{') != std::string::npos && line.find('{') == 0;
}

int processFile(const std::string& filename) {
    std::ifstream file(filename);
    std::vector<std::string> lines;
    std::string line;
    while (std::getline(file, line)) {
        lines.push_back(line);
    }
    file.close();

    std::vector<std::string> newLines;
    int changeCounter = 0;
    size_t i = 0;
    while (i < lines.size()) {
        if (i < lines.size() - 3 && checkFirstLine(lines[i]) && checkForContent(lines[i+1]) && checkForEndParen(lines[i+2]) && checkForFuncStart(lines[i+3])) {
            newLines.push_back(lines[i]);
            newLines.push_back(lines[i+1] + " " + lines[i+2]);
            newLines.push_back(lines[i+3]);
            i += 4;
            changeCounter++;
        } else if (i < lines.size() - 4 && checkFirstLine(lines[i]) && checkForContent(lines[i+1]) && checkForParen(lines[i+2]) && checkForEndParen(lines[i+3]) && checkForFuncStart(lines[i+4])) {
            newLines.push_back(lines[i]);
            newLines.push_back(lines[i+1] + " " + lines[i+2]);
            newLines.push_back(lines[i+3]);
            newLines.push_back(lines[i+4]);
            i += 5;
            changeCounter++;
        } else if (i < lines.size() - 5 && checkFirstLine(lines[i]) && checkForContent(lines[i+1]) && checkForParen(lines[i+2]) && checkForContent(lines[i+3]) && checkForEndParen(lines[i+4]) && checkForFuncStart(lines[i+5])) {
            std::cout << "Found function with three lines of arguments, see: " << lines[i+1] << " " << lines[i+2] << " " << lines[i+3] << " " << lines[i+4] << std::endl;
            newLines.push_back(lines[i]);
            i++;
        } else {
            newLines.push_back(lines[i]);
            i++;
        }
    }

    std::ofstream outFile(filename);
    for (const auto& newLine : newLines) {
        outFile << newLine << "\n";
    }
    outFile.close();

    return changeCounter;
}

int main() {
    std::string path = "./"; // Current directory
    int totalChangeCount = 0;
    for (const auto& entry : fs::directory_iterator(path)) {
        if (entry.path().extension() == ".c") {
            std::cout << "Processing file: " << entry.path() << std::endl;
            int fileChangeCount = processFile(entry.path());
            std::cout << "Done! Changes done: " << fileChangeCount << std::endl;
            totalChangeCount += fileChangeCount;
        }
    }

    std::cout << "Done processing files! Total changes done: " << totalChangeCount << std::endl;
    return 0;
}
